packages/flutter/lib/src/widgets/window.dart

```diff
@@ -3,169 +3,331 @@
+/// listening to changes on the window.
+abstract class WindowController with ChangeNotifier {
+  FlutterView? _view;
+
+  /// The ID of the view used for this window, which is unique to each window.
+  FlutterView? get view => _view;
+  set view(FlutterView? value) {
+    _view = value;
+    notifyListeners();
+  }
 
-  /// Defines a dialog [Window]
-  dialog,
+  Size? _size;
 
-  /// Defines a satellite [Window]
-  satellite,
+  /// The current size of the window. This may differ from the requested size.
+  Size? get size => _size;
+  set size(Size? value) {
+    _size = value;
+    notifyListeners();
+  }
 
-  /// Defines a popup [Window]
-  popup,
+  int? _parentViewId;
 
-  /// Defines a tooltip
-  tip,
-}
+  /// The ID of the parent in which this rendered, if any.
+  int? get parentViewId => _parentViewId;
+  set parentViewId(int? value) {
+    _parentViewId = value;
+    notifyListeners();
+  }
 
-class WindowChangedEvent {
-  WindowChangedEvent({this.size});
-  Size? size;
+  /// The archetype of the window.
+  WindowArchetype get type;
+
+  /// Destroys this window.
+  Future<void> destroy() async {
+    if (view == null) {
+      return;
+    }
+
+    return destroyWindow(view!.viewId);
+  }
 }
 
-class _WindowMetadata {
-  _WindowMetadata(
-      {required this.flView,
-      required this.archetype,
-      required this.size,
-      this.parent});
+/// Provided to [RegularWindow]. Allows the user to listen on changes
+/// to a regular window and modify the window.
+class RegularWindowController extends WindowController {
+  @override
+  WindowArchetype get type => WindowArchetype.regular;
 
-  final FlutterView flView;
-  final WindowArchetype archetype;
-  final Size size;
-  final Window? parent;
+  /// Modify the properties of the window.
+  Future<void> modify({Size? size}) {
+    throw UnimplementedError();
+  }
 }
 
-/// Defines a [Window] created by the application. To use [Window]s, you must wrap
-/// your application in the [MultiWindowApp] widget. New [Window]s are created via
-/// global functions like [createRegular] and [createPopup].
-abstract class Window {
-  /// [view] the underlying [FlutterView]
-  /// [builder] render function containing the content of this [Window]
-  /// [size] initial [Size] of the [Window]
-  /// [parent] the parent of this window, if any
-  Window(
-      {required this.view,
-      required this.builder,
-      required this.size,
-      this.parent});
+/// A widget that creates a regular window. This content of this window is
+/// rendered into a [View], meaning that this widget must be rendered into
+/// either a [ViewAnchor] or a [ViewCollection].
+class RegularWindow extends StatefulWidget {
+  /// Creates a regular window widget
+  const RegularWindow(
+      {this.controller,
+      this.onDestroyed,
+      this.onError,
+      super.key,
+      required Size preferredSize,
+      required this.child})
+      : _preferredSize = preferredSize;
```

Comment:
Is it possible to use `this.preferredSize` directly instead of using a separate `_preferredSize` field?

```dart
const RegularWindow({
  super.key,
  this.controller,
  this.onDestroyed,
  this.onError,
  required this.preferredSize,
  required this.child,
});
```